File: modules/document/src/main/java/org/xsslang/xide/document/LineMap.java
/*
 * SPDX-FileCopyrightText: 2026 Leitwolf <xs-lang.chess031@slmails.com>
 * SPDX-License-Identifier: MPL-2.0
 */

package org.xsslang.xide.document;

import java.util.Arrays;

/**
 * Maps offsets to line positions without converting Java's native UTF-16 representation.
 */
public final class LineMap {
    private final int[] lineStarts;
    private final int[] lineEnds;
    private final int textLength;

    private LineMap(int[] lineStarts, int[] lineEnds, int textLength) {
        this.lineStarts = lineStarts;
        this.lineEnds = lineEnds;
        this.textLength = textLength;
    }

    /**
     * Scans text into an immutable line map.
     *
     * @param text source text
     * @return scanned line map
     */
    public static LineMap of(String text) {
        int[] starts = new int[Math.max(8, text.length() / 16)];
        int[] ends = new int[starts.length];
        int lineCount = 0;
        int lineStart = 0;

        for (int offset = 0; offset < text.length(); offset++) {
            char current = text.charAt(offset);
            if (current != '\r' && current != '\n') {
                continue;
            }
            if (lineCount == starts.length) {
                starts = Arrays.copyOf(starts, starts.length * 2);
                ends = Arrays.copyOf(ends, ends.length * 2);
            }
            starts[lineCount] = lineStart;
            ends[lineCount++] = offset;
            if (current == '\r' && offset + 1 < text.length() && text.charAt(offset + 1) == '\n') {
                offset++;
            }
            lineStart = offset + 1;
        }

        if (lineCount == starts.length) {
            starts = Arrays.copyOf(starts, starts.length + 1);
            ends = Arrays.copyOf(ends, ends.length + 1);
        }
        starts[lineCount] = lineStart;
        ends[lineCount++] = text.length();
        return new LineMap(Arrays.copyOf(starts, lineCount), Arrays.copyOf(ends, lineCount), text.length());
    }

    /**
     * Returns the number of logical lines, including the final empty line.
     *
     * @return line count
     */
    public int lineCount() {
        return lineStarts.length;
    }

    /**
     * Converts a line and column into a UTF-16 offset.
     *
     * @param position position to convert
     * @return UTF-16 offset
     */
    public int offsetAt(TextPosition position) {
        if (position.line() >= lineStarts.length) {
            throw new IndexOutOfBoundsException("line is outside the document");
        }
        int lineLength = lineEnds[position.line()] - lineStarts[position.line()];
        if (position.column() > lineLength) {
            throw new IndexOutOfBoundsException("column is outside the line");
        }
        return lineStarts[position.line()] + position.column();
    }

    /**
     * Converts a UTF-16 offset into a line and column.
     *
     * @param offset offset to convert
     * @return line and UTF-16 column
     */
    public TextPosition positionAt(int offset) {
        if (offset < 0 || offset > textLength) {
            throw new IndexOutOfBoundsException("offset is outside the document");
        }
        int search = Arrays.binarySearch(lineStarts, offset);
        int line = search >= 0 ? search : -search - 2;
        int column = Math.min(offset, lineEnds[line]) - lineStarts[line];
        return new TextPosition(line, column);
    }
}

File: modules/document/src/main/kotlin/com/progmasoft/xide/document/LineMap.kt
/*
 * SPDX-FileCopyrightText: 2026 Progmasoft <support@progmasoft.com>
 * SPDX-License-Identifier: MPL-2.0 WITH AdditionRef-Progmasoft-Exception-1.0
 */

package com.progmasoft.xide.document

/** Maps offsets without converting the JVM's native UTF-16 string representation. */
class LineMap private constructor(
  private val lineStarts: IntArray,
  private val lineEnds: IntArray,
  private val textLength: Int,
) {
  val lineCount: Int
    get() = lineStarts.size

  fun offsetAt(position: TextPosition): Int {
    if (position.line !in lineStarts.indices) {
      throw IndexOutOfBoundsException("line is outside the document")
    }
    val lineLength = lineEnds[position.line] - lineStarts[position.line]
    if (position.column > lineLength) {
      throw IndexOutOfBoundsException("column is outside the line")
    }
    return lineStarts[position.line] + position.column
  }

  fun positionAt(offset: Int): TextPosition {
    if (offset !in 0..textLength) {
      throw IndexOutOfBoundsException("offset is outside the document")
    }
    val search = lineStarts.binarySearch(offset)
    val line = if (search >= 0) search else -search - 2
    val column = minOf(offset, lineEnds[line]) - lineStarts[line]
    return TextPosition(line, column)
  }

  companion object {
    fun of(text: String): LineMap {
      val starts = mutableListOf<Int>()
      val ends = mutableListOf<Int>()
      var lineStart = 0
      var offset = 0
      while (offset < text.length) {
        val current = text[offset]
        if (current != '\r' && current != '\n') {
          offset++
          continue
        }
        starts += lineStart
        ends += offset
        if (current == '\r' && offset + 1 < text.length && text[offset + 1] == '\n') {
          offset++
        }
        lineStart = ++offset
      }
      starts += lineStart
      ends += text.length
      return LineMap(starts.toIntArray(), ends.toIntArray(), text.length)
    }
  }
}
